pyomo/contrib/piecewise/transform/incremental.py

```diff
@@ -0,0 +1,203 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2024
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.contrib.fbbt.fbbt import compute_bounds_on_expr
+from pyomo.contrib.piecewise.transform.piecewise_linear_transformation_base import (
+    PiecewiseLinearTransformationBase,
+)
+from pyomo.contrib.piecewise.triangulations import Triangulation
+from pyomo.core import (
+    Constraint,
+    Binary,
+    NonNegativeIntegers,
+    Suffix,
+    Var,
+    RangeSet,
+    Param,
+)
+from pyomo.core.base import TransformationFactory
+from pyomo.gdp import Disjunct, Disjunction
+from pyomo.common.errors import DeveloperError
+from pyomo.core.expr.visitor import SimpleExpressionVisitor
+from pyomo.core.expr.current import identify_components
+from math import ceil, log2
+import logging
+
+
+@TransformationFactory.register(
+    "contrib.piecewise.incremental",
+    doc="""
+    The incremental MIP formulation of a piecewise-linear function, as described
+    by [1]. To work in the multivariate case, the underlying triangulation must 
+    satisfy these properties:
+     (1) The simplices are ordered T_1, ..., T_N such that T_i has nonempty intersection
+         with T_{i+1}. It doesn't have to be a whole face; just a vertex is enough.
+     (2) On each simplex T_i, the vertices are ordered T_i^1, ..., T_i^n such
+         that T_i^n = T_{i+1}^1
+    In Pyomo, the Triangulation.OrderedJ1 triangulation is compatible with this
+    transformation.
+
+    References
+    ----------
+    [1] J.P. Vielma, S. Ahmed, and G. Nemhauser, "Mixed-integer models
+        for nonseparable piecewise-linear optimization: unifying framework
+        and extensions," Operations Research, vol. 58, no. 2, pp. 305-315,
+        2010.
+    """,
+)
+class IncrementalMIPTransformation(PiecewiseLinearTransformationBase):
+
+    CONFIG = PiecewiseLinearTransformationBase.CONFIG()
+    _transformation_name = "pw_linear_incremental"
+
+    # Implement to use PiecewiseLinearTransformationBase. This function returns the Var
+    # that replaces the transformed piecewise linear expr
+    def _transform_pw_linear_expr(self, pw_expr, pw_linear_func, transformation_block):
+        if pw_linear_func.triangulation not in (
+            Triangulation.OrderedJ1,
+            Triangulation.AssumeValid,
```

Comment:
Isn't `AssumeValid` a misnomer if you aren't trusting it here? I think trusting it makes more sense than the override.

Reply:
I am trusting it. It's a whitelist, we error out if they pass anything besides an OrderedJ1 or AssumeValid.